pkg/bgpv1/agent/controller.go

```
@@ -391,3 +395,20 @@ func (c *Controller) Reconcile(ctx context.Context) error {
 func (c *Controller) FullWithdrawal(ctx context.Context) {
 	_ = c.BGPMgr.ConfigurePeers(ctx, nil, nil) // cannot fail, no need for error handling
 }
+
+// applyPolicyDefaults applies default values on the CiliumBGPPeeringPolicy.
+func (c *Controller) applyPolicyDefaults(policy *v2alpha1api.CiliumBGPPeeringPolicy) {
+	for _, r := range policy.Spec.VirtualRouters {
+		for _, n := range r.Neighbors {
+			if n.ConnectRetryTime.Duration == 0 {
+				n.ConnectRetryTime.Duration = types.DefaultBGPConnectRetryTime
+			}
+			if n.HoldTime.Duration == 0 {
+				n.HoldTime.Duration = types.DefaultBGPHoldTime
+			}
+			if n.KeepAliveTime.Duration == 0 {
+				n.KeepAliveTime.Duration = n.HoldTime.Duration / 3
+			}
```

Have you already checked if any of these parameters have a special meaning for 0? For example, RFC4271 Sec 4.4 says that hold time = 0 has a special meaning that disables keepalive. I don't know how practical it is to use that, but how about making the CRD field nullable and checking whether it is nil here instead of checking zero?

---

@YutaroHayakawa You are right about the `0` hold time parameter - it does have special meaning per RFC. However, in GoBGP the API of all timer params is in `uint64` and if `0` is passed, they are defaulted to non-zero values, including the hold time (see [the code](https://github.com/osrg/gobgp/blob/2fbfee5f2748e387e4442164760a6c56c1cc0578/internal/pkg/config/default.go#L117-L125)). So in fact, if we allowed `nil` value in our API and would not default `0`, it would be still defaulted on the GoBGP level to non-0, which I think could be misleading to the user - that's why I opted to defaulting exactly the same way as in GoBGP...

In general I agree that distinguishing between `0` and not-set value makes sense, so I am not sure myself how much we should tie our API to the GoBGP behavior here. Maybe if we potentially plan for other BGP backends that may properly support setting `0` hold time value (e.g. FRR seems to support it), allowing `nil` and not defaulting on `0` + documenting the behavior of GoBGP backend would make sense?

---

> However, in GoBGP the API of all timer params is in uint64 and if 0 is passed, they are defaulted to non-zero values, including the hold time (see [the code](https://github.com/osrg/gobgp/blob/2fbfee5f2748e387e4442164760a6c56c1cc0578/internal/pkg/config/default.go#L117-L125)).

Ohh, I see. Thanks for the clarification. Alright, then, I think we don't have to make that configurable. Even if we want to add a feature to set it to zero, later on, we can add one more boolean flag.